lidofinance/anchor-collateral-steth
# Integration with Wormhole bridge

## Preface
This PR would introduce integration with [Wormhole v2 bridge](https://github.com/certusone/wormhole/tree/dev.v2) respecting the [specification draft](https://hackmd.io/@blitz-1306/B1oUJS6rt).

## Changes
- [x] Added `requirements-dev.txt` to simplify work environment installation.
- [x] Introduced on-chain contract addresses to constants.
- [x] Introduced `wormhole_token_bridge` state vaiable and paramater in constructor (to use mocks).
- [x] Renamed method `_lockAssets()` to `_transfer_asset()`, added parameter `_extra_data` to handle `nonce` and `arbiter_fee` parsing from custom payload of bytes.
- [x] Changed raw call to use Wormhole v2 [`Bridge.transferTokens()` method signature and arguments](https://github.com/certusone/wormhole/blob/4c85daa53118a1ceff90949c27ca03bf4ac035c9/ethereum/contracts/bridge/Bridge.sol#L93).
- [x] Tweaked method `adjust_amount()` to follow [decimals normalization logic in bridge](https://github.com/certusone/wormhole/blob/4c85daa53118a1ceff90949c27ca03bf4ac035c9/ethereum/contracts/bridge/Bridge.sol#L112-L120).
- [x] Tweaked methods `forward_beth()` and `forward_ust()` to respect changes.
- [x] Introduced `tests/test_connector_wormhole.py` unit test suite.
- [x] Introduced `scripts/wormhole_connector_check.py` script to check bringe switching flow.
- [x] Introduced `scripts/deploy_wormhole_connector.py` script for simplified bridge connector deployment and changing it in vault.

## Notes
1. Starting 64 bytes of `_extra_data` are reserved to pass `nonce` and `arbiter_fee` values.
2. `nonce` is restricted to `uint32` type range boudaries (4 bytes maximum) for compatibility with [`Bridge.transferTokens()` method signature](https://github.com/certusone/wormhole/blob/4c85daa53118a1ceff90949c27ca03bf4ac035c9/ethereum/contracts/bridge/Bridge.sol#L93).

## Testing
1. Setup environment:
    ```bash
    # Install Ganache CLI:
    npm install -g ganache-cli

    # Setup virtual environment for Python:
    virtualenv venv --python=python3.8

    # Activate virtual environment:
    source venv/bin/activate

    # Install dependencies:
    pip install -r requirements-dev.txt
    ```
2. Configure chain fork by setting `fork` in `brownie-config.yaml`.
3. Run tests:
    ```bash
    brownie test
    # or particularly
    brownie test tests/test_connector_wormhole.py
    ```
4. Run happy path script:
    ```bash
    brownie run scripts/wormhole_connector_check.py
    ```
5. Run standalone script for bridge connector deployment and setting in `AnchorVault` contract:
    ```bash
    export DEPLOYER=<deployer account address>
    export CHANGER=<connector changer account address>

    brownie run scripts/deploy_wormhole_connector.py
    ```
    P.S.: Ensure that specified accounts have enough funds on their balance.

Regards.
